### ISSUE TITLE
[🐛BUG] Invalid 'neg_sampling' parameter

### ISSUE THREAD

**Issue by Tokkiu** [RUCAIBox/RecBole#1249]
**Describe the bug**
I specify the 'neg_sampling' as '--neg_sampling={'uniform':2}'. And I can see the parameter in training log.
<img width="253" alt="Screenshot 2022-04-14 at 9 05 44 PM" src="https://user-images.githubusercontent.com/13414571/163396778-79d9dbc9-ef69-46c5-b5c6-e85dd1af93c0.png">
However, I found my neg_items.shape is [B] which variable is assigned by 
```
neg_items = interaction[self.NEG_ITEM_ID]
```
My expected neg_items.shape should be [B, 2]

**To Reproduce**
```
Python run_recbole.py --model=SASRec --neg_sampling={'uniform':2} --loss_type=BPR
```

**Expected behavior**
neg_items.shape = [B, 2]

**Comment by Sherry-XLL** [comment]
Hello, @Tokkiu , `neg_items` is a one-dimensional array regardless of the number of negative samples. 

If `train_batch_size=B`, and the number of negative samples equals `K`, `neg_items.shape` should be `[B * K]` instead of `[B, K]`. For details, you can refer to the codes and comments in the `sampler` section as follows.

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/recbole/sampler/sampler.py#L265-L279

If you want to get a two-dimensional array, transform the array through `reshape` or `view`.

Thanks for your attention to RecBole!

**Comment by Sherry-XLL** [comment]
@Tokkiu Sorry, there is something wrong with my statement just now. In order to control the usage of memory, RecBole will always ensure that the whole size of each batch does not exceed `train_batch_size`, while the real batch size `B` is `self.step = max(self.config['train_batch_size'] // self.times, 1)`. For the result of every batch, we permit that every positive interaction and its negative interaction must be in the same batch. 

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/recbole/data/dataloader/general_dataloader.py#L40-L46

As a result, `neg_items.shape` is `[B * 2]`, while `B` is `train_batch_size // 2` in your case.

**Comment by Tokkiu** [comment]
@Sherry-XLL Thanks for your information.
So in your opinion, the Recbole will produce two pairs of <pos_item, neg_item>, right?
For example, when I specify the neg_sampling is 2. Then for each postive item i (term it as i_p), I can have my batch like this:
[
 ...
 [i_p, i_n_1],
 [i_p, i_n_2],
 ...
]
Then when I calculate the loss, the loss for i_p will be calculated twice.
If I want to calculate the BCE for item i, I will get 
loss_i = BCE(i_p, i_n_1) + BCE(i_p, i_n_2)

Am I right about this calculation process?

**Comment by Sherry-XLL** [comment]
@Tokkiu The processing flow is different for different types of loss functions. 

BPR (Bayesian Personalized Ranking) is a pairwise loss, so each negative sample is paired with a positive sample. 

BCE (Binary Cross Entropy) is a pointwise loss, and labels (0/1) determine the positive or negative samples. In this case, the positive samples will not be repeated.

What's more, negative sampling is not required in CE loss of SASRec.

For more details, you can refer to the code in dataloader.

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/recbole/data/dataloader/abstract_dataloader.py#L182-L198

**Comment by Tokkiu** [comment]
@Sherry-XLL Thanks for your kind notice, you solved my question well. I will close this issue asap.
